### PR TITLE
net: remove -banscore configuration option

### PR DESCRIPTION
**Pull request by jonatack**
per https://github.com/bitcoin/bitcoin/pull/19219#issuecomment-652684340, https://github.com/bitcoin/bitcoin/pull/19219#discussion_r443074487 and https://github.com/bitcoin/bitcoin/pull/19219#issuecomment-652699592. Edit: now split into 3 PRs:
- net: remove -banscore configuration option (this PR)
- rpc: deprecate banscore field in getpeerinfo (#19469)
- gui, doc, doc: no longer display "Ban Score" in GUI Peers window + test/doc fixups (#19512)

### CODE REVIEW

**Comment by jonatack** [comment]
If it's simpler to keep these (straightforward) changes here in one PR, even better.

**Comment by jonatack** [comment]
Now that #19469 has been merged, please review this companion PR. It's a small change that should be easy to review.

File: doc/release-notes-19464.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Updated settings
+----------------
+
+- The `-banscore` configuration option, which modified the default threshold for
+  disconnecting and discouraging misbehaving peers, has been removed as part of
+  changes in this release to the handling of misbehaving peers. Refer to the
+  section, "Changes regarding misbehaving peers", for details. (#19464)
```
Comment (MarcoFalke) [inline doc/release-notes-19464.md:7]: Mind adding this to the main release notes file? There shouldn't be any conflicts and this is not for backport either.

**Review by maflcko — COMMENTED** [review]
Concept aCK

**Comment by fjahr** [comment]
Concept ACK but the commits are cluttered up with quite a bit of edits that produce unnecessary noise.

File: doc/release-notes-19464.md (hunk 1 above)
Comment (jonatack) [reply doc/release-notes-19464.md]: Done

**Comment by jonatack** [comment]
Thanks @MarcoFalke and @fjahr for reviewing! Updated with your feedback per `git range-diff 505b4ed f926882 b71eac1` for easy re-ACK.

File: src/net_processing.h [hunk 2]
```diff
@@ -23,6 +23,7 @@ static const unsigned int DEFAULT_MAX_ORPHAN_TRANSACTIONS = 100;
 static const unsigned int DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN = 100;
 static const bool DEFAULT_PEERBLOOMFILTERS = false;
 static const bool DEFAULT_PEERBLOCKFILTERS = false;
+static const int BANSCORE_THRESHOLD = 100;
```
Comment (MarcoFalke) [inline src/net_processing.h:26]: can be moved to the cpp file and made constexpr?

**Review by maflcko — APPROVED** [review]
Do the unit and functional tests pass on ff43c0c49eb087d27cd6ecf6b30fe6f6f96df650

If not, the following two commits should be squashed into it. Or maybe just squash all commit? :man_shrugging: 

Review-only ACK

**Comment by jonatack** [comment]
> Do the unit and functional tests pass on [ff43c0c](https://github.com/bitcoin/bitcoin/commit/ff43c0c49eb087d27cd6ecf6b30fe6f6f96df650)

The tests had to be updated for the config removal.

**Comment by jonatack** [comment]
> If not, the following two commits should be squashed into it. Or maybe just squash all commit? man_shrugging
> 
> Review-only ACK

Thanks for ACKing. I took the time to made atomic commits for different actions, step by step, as seen in other contributors' changesets and as described in `CONTRIBUTING.md` and in the wikipedia link it contains.
```md
In general, [commits should be atomic](https://en.wikipedia.org/wiki/Atomic_commit#Atomic_commit_convention)
and diffs should be easy to read. For this reason, do not mix any formatting
fixes or code moves with actual code changes.
```

**Comment by jonatack** [comment]
I also tried to make these changes clean and minimal without noise :man_shrugging:

File: src/net_processing.h (hunk 2 above)
Comment (jonatack) [reply src/net_processing.h:26]: > can be moved to the cpp file and made constexpr?

good idea

**Comment by MarcoFalke** [comment]
Please see the second section of https://www.codewithjason.com/atomic-commits-testing/

**Comment by jonatack** [comment]
> Please see the second section of https://www.codewithjason.com/atomic-commits-testing/

Yes, I agree with this and it has been my standard practice for decades (except for the first decade which was working in assembly language). But AFAIK it's not what is requested in this project's documentation, and not the observed practice -- I had to unlearn my practices to adapt to it.

**Comment by MarcoFalke** [comment]
>  But AFAIK it's not what is requested in this project's documentation

Ok, then our documentation is wrong and should be fixed.

**Comment by jonatack** [comment]
I'll make the commits hygienic.

File: src/net_processing.h (hunk 2 above)
Comment (jonatack) [reply src/net_processing.h]: Done, then moved it back to the header file to be able to use it in the unit tests. This allowed for some nice improvements there.

**Comment by jonatack** [comment]
Rebased to see if the latest merge on master solves the travis issues.

File: doc/release-notes.md [hunk 3]
```diff
@@ -99,6 +99,11 @@ Build System
 Updated settings
 ----------------
 
+- The `-banscore` configuration option, which modified the default threshold for
+  disconnecting and discouraging misbehaving peers, has been removed as part of
+  changes in this release to the handling of misbehaving peers. Refer to the
+  section, "Changes regarding misbehaving peers", for details. (#19464)
+
```
Comment (vasild) [inline doc/release-notes.md:106]: Shouldn't this be in `/doc/release-notes-19464.md`? `doc/developer-notes.md` says:
> Release notes should be added to a PR-specific release note file at
`/doc/release-notes-<PR number>.md`

File: src/net_processing.h [hunk 4]
```diff
@@ -23,6 +23,8 @@ static const unsigned int DEFAULT_MAX_ORPHAN_TRANSACTIONS = 100;
 static const unsigned int DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN = 100;
 static const bool DEFAULT_PEERBLOOMFILTERS = false;
 static const bool DEFAULT_PEERBLOCKFILTERS = false;
+/** Threshold for marking a node to be discouraged, e.g. disconnected and added to the discouragement filter. */
+static const int DISCOURAGEMENT_THRESHOLD{100};
```
Comment (vasild) [inline src/net_processing.h:27]: Maybe worth mentioning what is the unit of this?

**Review by vasild — COMMENTED** [review]
Should `doc/man/bitcoin-qt.1` and `doc/man/bitcoind.1` be updated in this PR (they contain references to `-banscore`)?

If this PR gets merged now, it means that Bitcoin Core 0.21 will not recognize the `-banscore` command line option, is this the intention?

File: doc/release-notes.md (hunk 3 above)
Comment (jonatack) [reply doc/release-notes.md:106]: See https://github.com/bitcoin/bitcoin/pull/19464#discussion_r452835586

File: src/net_processing.h (hunk 4 above)
Comment (jonatack) [reply src/net_processing.h:27]: What unit would you suggest? It's just a score, and expected to be removed (see the links in the PR description).

**Comment by jonatack** [comment]
Thanks for reviewing, @vasild!

> Should `doc/man/bitcoin-qt.1` and `doc/man/bitcoind.1` be updated in this PR (they contain references to `-banscore`)?

AFAIK the manpages are auto-generated; we don't edit them in PRs.
 
> If this PR gets merged now, it means that Bitcoin Core 0.21 will not recognize the `-banscore` command line option, is this the intention?

Correct. See the links in the PR description for intention and reasoning behind the removal.

AFAIK there isn't a deprecation process here like for the RPC API, so either it's removed as proposed here, or we add a deprecation warning in the help (which everyone will ignore until the option is removed) for release or two, I suppose.

**Comment by MarcoFalke** [comment]
Ideally everything would go in one swish (including state->nMisbehavior), but that can't happen because it is still needed for RPC. This is a necessary first step.


review ACK 06059b0c2a6c2db70c87a7715f8a344a13400fa1 📙

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 06059b0c2a6c2db70c87a7715f8a344a13400fa1 📙
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjlzAv+Mu1M+kc3EzVlpVHUNyR9vN531cYpgnWG+gct6oTejWWTBb8snhDrkAID
+J4QbpAVd7LBg6CV2NJAl2I0/mTbcSuA66z9DhPwhptKnzocbxf4qXvj5sBnIxg8
MOOcRAz5mGVqTom64t2EOme4qKnGvHQROY4srg/Iaf87+Xeur4qBFLuwezGOG9VH
xEXSJlaZ0MM9zhLfuJpAIOE0NYiAwPRWnihCofbspg7Mv3duXiOna4gsF7TXV3LA
uR/tLwZaZJkQRugKA0Dd5tBFu4+tPqZQ/blyFyGdbSunHVIJnyvLwA7T/nnz0YSO
vcgJn8nFYO/MObhyBgWST0Bv6I/MsaB9it7cQnbXQjDtAAd4oUqeOzH5gbZYRSxy
MqSRf/4eKcefvx0o5oIuwlva4HjUIwm68t2rn8lK/nSdERRGILCXE//SaIT9NUCc
nbOIVLiQvVembkFWEZMWatZm9ty93a5ImbAY7YvyCBSK6/1vqtUClIXAHLU8+6oM
2HQc/Pec
=/Sro
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `7468957afe68f0b446600d6fe8fc55ac3e9656a9dfb3300c2ac5b7e9c089a92c  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401087468957afe68f0b446600d6fe8fc55ac3e9656a9dfb3300c2ac5b7e9c089a92cf0101e8fd2a3c3866f5306e4b3712fb87e8708fff010576ab20fadf2707d0222090a9ae1059008f1045f0c9990f0085aa341ffe404d7760083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01095007a5afedd8e45c79958cb3aae9a6808f1045f0c998ff008e3d39de9ba2072540083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0109c44ab67a663b87152e54def26f23ac608f1045f0c9990f00877a380dc6bc23b8e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010e2cacc83a5254cecdc16fa7a68d3cdc408f1045f0c9990f00826a2c7ca04143fa60083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

**Review by vasild — APPROVED** [review]
ACK 06059b0c

File: test/functional/p2p_leak.py [hunk 5]
```diff
@@ -70,7 +70,7 @@ class CNodeNoVersionBan(CLazyNode):
     # NOTE: implementation-specific check here. Remove if bitcoind ban behavior changes
     def on_open(self):
         super().on_open()
-        for i in range(banscore):
+        for _ in range(DISCOURAGEMENT_THRESHOLD):
```
Comment (practicalswift) [inline test/functional/p2p_leak.py:73]: Why the change from 10 to 100 iterations here? :)

File: src/test/denialofservice_tests.cpp [hunk 6]
```diff
@@ -288,7 +287,7 @@ BOOST_AUTO_TEST_CASE(DoS_banscore)
     dummyNode1.fSuccessfullyConnected = true;
     {
         LOCK(cs_main);
-        Misbehaving(dummyNode1.GetId(), 100);
+        Misbehaving(dummyNode1.GetId(), DISCOURAGEMENT_THRESHOLD - 11);
```
Comment (practicalswift) [inline src/test/denialofservice_tests.cpp:290]: Why the change from `100` to `89` here? :)

**Review by practicalswift — COMMENTED** [review]
Concept ACK: giving users knobs without a recommendation about how/when to use them is not friendly :)

File: test/functional/p2p_leak.py (hunk 5 above)
Comment (jonatack) [reply test/functional/p2p_leak.py:73]: Because we can no longer use `-banscore` to change the default threshold (see removed line in test: `self.extra_args = [['-banscore=' + str(banscore)]]`), so we have to use `DISCOURAGEMENT_THRESHOLD` here for the number of veracks sans version to send.

Note to self: the words "ban" in this functional test should be updated to "discourage".

File: src/test/denialofservice_tests.cpp (hunk 6 above)
Comment (jonatack) [reply src/test/denialofservice_tests.cpp:290]: See the two removed `-banscore` lines of this test, particularly the first one: `gArgs.ForceSetArg("-banscore", "111"); // because 11 is my favorite number`.

**Review by maflcko — COMMENTED** [review]
Oh, I see this section won't end up in the 0.21 release notes, so might want to clarify at some point

File: doc/release-notes.md [hunk 7]
```diff
@@ -99,6 +99,11 @@ Build System
 Updated settings
 ----------------
 
+- The `-banscore` configuration option, which modified the default threshold for
+  disconnecting and discouraging misbehaving peers, has been removed as part of
+  changes in this release to the handling of misbehaving peers. Refer to the
+  section, "Changes regarding misbehaving peers", for details. (#19464)
```
[the remark this answers is not in the document]
Comment (jonatack) [reply doc/release-notes.md:105]: Oh, good catch.

[the remark this answers is not in the document]
Comment (jonatack) [reply doc/release-notes.md:105]: Will touch it up in the GUI PR to remove the banscore field, mentioned in the PR description.

[the remark this answers is not in the document]
Comment (jonatack) [reply doc/release-notes.md:105]: Done, thanks again for seeing that.

**Comment by jnewbery** [comment]
post-merge ACK.

The test `BOOST_AUTO_TEST_CASE(DoS_banscore)` should be removed entirely. It only existed to test manually testing the `-banscore` config. Nothing else in the test is untested by the previous `DoS_banning` test.

**Comment by jonatack** [comment]
> post-merge ACK.
> 
> The test `BOOST_AUTO_TEST_CASE(DoS_banscore)` should be removed entirely. It only existed to test manually testing the `-banscore` config. Nothing else in the test is untested by the previous `DoS_banning` test.

Good point @jnewbery, I'll look at tucking that removal into the test change commit in #19512.

**Comment by jonatack** [comment]
> post-merge ACK.
> 
> The test `BOOST_AUTO_TEST_CASE(DoS_banscore)` should be removed entirely. It only existed to test manually testing the `-banscore` config. Nothing else in the test is untested by the previous `DoS_banning` test.

Good point @jnewbery. Added the removal to the test change commit in #19512.
